perf: dialtone-icons tree shaking

juliodialpad opened the pull request:
# Dialtone-icons tree shaking

## Obligatory GIF (super important!)

![Obligatory GIF](https://media.giphy.com/media/f9SiPj5zu0HrXlR8zD/giphy.gif?cid=790b7611gatca0nu9epg4p1a70nf4y3ttbpvx3oq0v7m1mmq&ep=v1_gifs_search&rid=giphy.gif&ct=g)

## :hammer_and_wrench: Type Of Change

These types will increment the version number on release:

- [x] Performance Improvement

## :book: Jira Ticket

https://dialpad.atlassian.net/browse/DLT-1654

## :book: Description

- Added tree-shaking and dynamic import capabilities to dialtone-icons.
- Removed support for CJS, do we still need to support this? I don't think we're using that on any of our apps.
- Moved files to a workspace common directory.
- Added an automated index.js file (gitignored) to dialtone-icons vue2 and vue3 version to function as an entry point.

## :bulb: Context

We're improving our libraries to make them lightweight and tree-shakeable

## :pencil: Checklist

For all PRs:

- [x] I have reviewed my changes.
- [x] I have considered the performance impact of my change.

juliodialpad committed:
move components_list.cjs to workspace common

juliodialpad committed:
move dialtone-vue build docs to workspace scripts

braddialpad committed:
fix(emoji): dlt-1650 pad emoji unicode string with 0's (#219)

juliodialpad committed:
feat: dynamic import icons

ninamarina committed:
docs: add tokens keywords (#220)

juliodialpad committed:
remove duplicated icons

braddialpad commented:
> Removed support for CJS, do we still need to support this? I don't think we're using that on any of our apps.

You mean the dialtone cjs dist output right? not internal cjs usage?

Not totally sure if anyone is using it.. I'll check some of of our applications, but yeah it would be nice to remove.

yorbi-dp committed:
feat: added new headset icon and update listed icons (#224)

braddialpad commented on packages/dialtone-vue2/components/icon/icon.vue:
```diff
@@ -62,13 +61,25 @@ export default {
+      async handler (name) {
+        const icon = await import(`/node_modules/@dialpad/dialtone-icons/vue2/dist/components/${name}.js`);
+        this.icon = icon.default;
+      },
```
I remember trying to do a dynamic import within our library before and it errored out when they tried to use it on the application side. Our infrastructure is much different than back then though, so maybe it works now. Just something to look out for when we try this out.

juliodialpad replied:
Tested on multiple sample applications and working well so far so I don't expect any mayor issue while testing on product.

braddialpad replied:
k sounds good.

braddialpad commented on packages/dialtone-icons/package.json:
```diff
@@ -6,18 +6,17 @@
+    "./vue2/*": "./vue2/dist/components/*.js",
+    "./vue3/*": "./vue3/dist/components/*.js",
+    "./svg/*": "./dist/svg/*",
+    "./*.json": "./dist/*.json"
```
Should we still allow `"./*": "./dist/*"` ?

juliodialpad replied:
No, `"./svg/*"` and `"./*.json"` are covering the exports on dist and `"./vue2/*"` and `"./vue3/*"` are covering the tree-shake on vue 2 and vue 3 icon versions.

braddialpad commented on packages/dialtone-icons/package.json:
```diff
@@ -65,5 +64,6 @@
   ],
-  "license": "MIT"
+  "license": "MIT",
+  "sideEffects": false
```
Can we explain or document what sideEffects are?

juliodialpad replied:
This is a left-over for testing tree-shake on webpack.

I'm not sure this is working as expected, it's not well documented so I'll remove it as it does nothing right now.

braddialpad replied:
"A "side effect" is defined as code that performs a special behavior when imported, other than exposing one or more exports. An example of this are polyfills, which affect the global scope and usually do not provide an export."

hmm don't really think we have to worry about that.

braddialpad replied:
or maybe we do have to set it to false on all our projects or webpack will not tree shake them? kinda seems that way

braddialpad replied:
https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free

braddialpad replied:
Seems like dialtone-css needs to be sideEffects: true based on this, since it's imported globally and css-loader is used. https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free

braddialpad commented on packages/dialtone-icons/package.json:
```diff
@@ -6,18 +6,17 @@
+    "./vue2": "./vue2/dist/dialtone-icons.js",
+    "./vue3": "./vue3/dist/dialtone-icons.js",
+    "./vue2/*": "./vue2/dist/components/*.js",
+    "./vue3/*": "./vue3/dist/components/*.js",
```
Hmm so does this mean consumers have to import components like `import DtIcon from '@dialpad/dialtone/vue2/icon'` or the code splitting doesn't work? Maybe will require testing product side, could depend on how the bundler is configured?

juliodialpad replied:
It's not enforced, consumers can still import from `'@dialpad/dialtone/vue2'`.

Importing from `DtIcon from '@dialpad/dialtone/vue2/icon'` instead of `{ DtIcon } from '@dialpad/dialtone/vue2'` it's a way to enforce the tree-shake behavior regardless of the consumer bundler config.

But that's on the dialtone-vue side, here we're making available to import every icon individually, like `import IconAccessibility from '@dialpad/dialtone-icons/vue2/accesibility'`

braddialpad replied:
Okay sweet that's what I was hoping.

braddialpad commented on packages/dialtone-icons/vue2/vite.config.js:
```diff
@@ -1,26 +1,32 @@
 export default defineConfig({
   plugins: [vue()],
   build: {
+    outDir: 'dist',
```
dist is the default value isn't it?

juliodialpad replied:
Yeah, I'll remove it

sanjeevdialpad committed:
fix: warn console error cleanup (#221)

juliodialpad commented:
> > Removed support for CJS, do we still need to support this? I don't think we're using that on any of our apps.
> 
> You mean the dialtone cjs dist output right? not internal cjs usage?
> 
> Not totally sure if anyone is using it.. I'll check some of of our applications, but yeah it would be nice to remove.

Yes, just the cjs output, or maybe there's a way to just generate dialtone-icons.cjs, not all the exports, I'll check if it's possible, if not, I think it's ok to just remove cjs output.

yorbi-dp committed:
feat: adding new blog for Types update (#222)

juliodialpad committed:
fix issues

ninamarina committed:
Update COMMIT_CONVENTION.md

semantic-release-bot committed:
chore(release): dialtone-icons/v4.12.0

# [4.12.0](https://github.com/dialpad/dialtone/compare/dialtone-icons/v4.11.0...dialtone-icons/v4.12.0) (2024-03-22)

### Features

* added new headset icon and update listed icons ([#224](https://github.com/dialpad/dialtone/issues/224)) ([c1373ec](https://github.com/dialpad/dialtone/commit/c1373ecb15492c7d20f7c2453b3def39457408e2))

semantic-release-bot committed:
chore(release): dialtone-vue3/v3.116.1

## [3.116.1](https://github.com/dialpad/dialtone/compare/dialtone-vue3/v3.116.0...dialtone-vue3/v3.116.1) (2024-03-22)

### Bug Fixes

* **Emoji:** dlt-1650 pad emoji unicode string with 0's ([#219](https://github.com/dialpad/dialtone/issues/219)) ([1eb4906](https://github.com/dialpad/dialtone/commit/1eb4906894537ed7ace2d9cad2de0513ff15a7cb))
* warn console error cleanup ([#221](https://github.com/dialpad/dialtone/issues/221)) ([65bb319](https://github.com/dialpad/dialtone/commit/65bb31983f74feed2e6d4d03581be58e3631b458))

semantic-release-bot committed:
chore(release): dialtone-vue2/v2.123.1

## [2.123.1](https://github.com/dialpad/dialtone/compare/dialtone-vue2/v2.123.0...dialtone-vue2/v2.123.1) (2024-03-22)

### Bug Fixes

* **Emoji:** dlt-1650 pad emoji unicode string with 0's ([#219](https://github.com/dialpad/dialtone/issues/219)) ([1eb4906](https://github.com/dialpad/dialtone/commit/1eb4906894537ed7ace2d9cad2de0513ff15a7cb))
* warn console error cleanup ([#221](https://github.com/dialpad/dialtone/issues/221)) ([65bb319](https://github.com/dialpad/dialtone/commit/65bb31983f74feed2e6d4d03581be58e3631b458))

semantic-release-bot committed:
chore(release): dialtone/v9.24.0

# [9.24.0](https://github.com/dialpad/dialtone/compare/dialtone/v9.23.0...dialtone/v9.24.0) (2024-03-22)

### Bug Fixes

* **Emoji:** dlt-1650 pad emoji unicode string with 0's ([#219](https://github.com/dialpad/dialtone/issues/219)) ([1eb4906](https://github.com/dialpad/dialtone/commit/1eb4906894537ed7ace2d9cad2de0513ff15a7cb))
* warn console error cleanup ([#221](https://github.com/dialpad/dialtone/issues/221)) ([65bb319](https://github.com/dialpad/dialtone/commit/65bb31983f74feed2e6d4d03581be58e3631b458))

### Documentation

* add tokens keywords ([#220](https://github.com/dialpad/dialtone/issues/220)) ([20125f4](https://github.com/dialpad/dialtone/commit/20125f421352fe2ccef2f092bb16c4a279d76b31))
* fix search for tokens ([#218](https://github.com/dialpad/dialtone/issues/218)) ([ff46724](https://github.com/dialpad/dialtone/commit/ff46724983c3b720fc6f48e71cbda88795e048f5))

### Features

* added new headset icon and update listed icons ([#224](https://github.com/dialpad/dialtone/issues/224)) ([c1373ec](https://github.com/dialpad/dialtone/commit/c1373ecb15492c7d20f7c2453b3def39457408e2))
* adding new blog for Types update ([#222](https://github.com/dialpad/dialtone/issues/222)) ([73b40f3](https://github.com/dialpad/dialtone/commit/73b40f30a4c2fd3ed26b213ebdcfce7d37ee2e62))

braddialpad committed:
chore: lint keywords

juliodialpad committed:
refactor dialtone-icons to remove the build process

juliodialpad committed:
update icon component

juliodialpad committed:
externalize deps

juliodialpad committed:
fix issues

juliodialpad commented:
After a long investigation the whole weekend trying to async load and tree-shake icons. Here are my thoughts:

- When async loading the spacing is not reserved until the icon is fully loaded, in order to solve this, I used DtSkeleton to reserve the space until the icon fully loads. After adding this, noticed a downgrade on performance on the `icon-variants` story (maybe because it's loading all the icons at once?).
- Our current DtIcon implementation is not tree-shakeable and it will never be as long as were using a prop to determine which icon should render, we'll need to do a big migration if we want to be able to fully tree-shake our icons.
- Modified dialtone-icons output to be able to use every icon as standalone component, passing the same props as `DtIcon`, this will allow us to make the migration gradually while still supporting the usage of `dt-icon`.
- The bundle size and performance of the consumer applications can be improved amazingly if we do our library components isolated from each other as much as possible and make the right imports to avoid polluting the bundle with unused code.

Here are some screenshots of the comparison between bundle sizes on test applications using different configs.

## Webpack

### Using DtIcon component

-  Using `import { DtIcon } from '@dialpad/dialtone-vue';`

![Screenshot 2024-03-25 at 4 37 31 a m](https://github.com/dialpad/dialtone/assets/87546543/169ceddc-dbab-4784-ad1d-48cdecda7e90)
![Screenshot 2024-03-25 at 4 37 56 a m](https://github.com/dialpad/dialtone/assets/87546543/5f166e42-f82e-43ca-a557-96becb78959f)

- Using `import { DtIcon } from '@dialpad/dialtone-vue/components/icon';`

![Screenshot 2024-03-25 at 4 35 05 a m](https://github.com/dialpad/dialtone/assets/87546543/7912e52b-c5e3-4233-8d20-5f50ec459d2d)
![Screenshot 2024-03-25 at 4 36 02 a m](https://github.com/dialpad/dialtone/assets/87546543/5afedb66-274a-4df8-8686-1e8ed6a980ea)

### Using dialtone-icons directly (suggested migration)

- Using `import { DtIconAccessibility } from '@dialpad/dialtone-icons/vue2';`

![Screenshot 2024-03-25 at 4 43 46 a m](https://github.com/dialpad/dialtone/assets/87546543/730c5248-bdbb-47c8-99ae-fbaa3e430780)
![Screenshot 2024-03-25 at 4 45 22 a m](https://github.com/dialpad/dialtone/assets/87546543/aa188714-ab6c-4296-91be-479bcc0a7747)

- Using `import DtIconAccessibility from '@dialpad/dialtone-icons/vue2/accessibility';`

![Screenshot 2024-03-25 at 4 47 39 a m](https://github.com/dialpad/dialtone/assets/87546543/8da13cdb-7390-41ab-8914-44a45bee4118)
![Screenshot 2024-03-25 at 4 48 09 a m](https://github.com/dialpad/dialtone/assets/87546543/84825a58-aea2-405a-92d6-3f2b88377b80)

## Vite

### Using DtIcon component

-  Using `import { DtIcon } from '@dialpad/dialtone-vue';`

![Screenshot 2024-03-25 at 4 56 15 a m](https://github.com/dialpad/dialtone/assets/87546543/c7fdaea8-eaf0-4273-97bb-f22137a94350)
![Screenshot 2024-03-25 at 4 56 53 a m](https://github.com/dialpad/dialtone/assets/87546543/9ecc2ce6-bf2c-4380-8eec-7237740c15ad)

- Using `import { DtIcon } from '@dialpad/dialtone-vue/components/icon';`

![Screenshot 2024-03-25 at 4 57 29 a m](https://github.com/dialpad/dialtone/assets/87546543/49b3e166-aeba-46dd-bd2c-7ef55abe4610)
![Screenshot 2024-03-25 at 4 57 44 a m](https://github.com/dialpad/dialtone/assets/87546543/447dd97d-52a1-4db4-807b-af9c7ab1949a)

### Using dialtone-icons directly (suggested migration)

- Using `import { DtIconAccessibility } from '@dialpad/dialtone-icons/vue2';`

![Screenshot 2024-03-25 at 4 58 58 a m](https://github.com/dialpad/dialtone/assets/87546543/3be56d51-fb52-4575-87b8-dbc3266c39f8)
![Screenshot 2024-03-25 at 4 59 19 a m](https://github.com/dialpad/dialtone/assets/87546543/8a4700f2-ed78-4c31-bc8e-b55970d8a374)

- Using `import DtIconAccessibility from '@dialpad/dialtone-icons/vue2/accessibility';`

![Screenshot 2024-03-25 at 4 59 58 a m](https://github.com/dialpad/dialtone/assets/87546543/4db6f534-4a41-4ad0-95f2-49af03bbf1ea)
![Screenshot 2024-03-25 at 5 00 11 a m](https://github.com/dialpad/dialtone/assets/87546543/540890eb-eb59-4dfa-ba8f-78a73242f213)

bianca-artola-dialpad committed:
fix(message-input): send complete files on add media (#230)

bianca-artola-dialpad committed:
fix(message-input): update mention suggestions rendering (#229)

ninamarina committed:
chore: DLT-1413 add jira id to commit lint (#228)

juliodialpad committed:
final version

juliodialpad committed:
Merge branch 'staging' into icons-tree-shaking

juliodialpad committed:
fix missing export

juliodialpad committed:
fix missing export

juliodialpad committed:
fix missing export

juliodialpad committed:
fix issues

juliodialpad committed:
fix spacing issues

juliodialpad commented:
Visual tests are marking all the snapshots as "new" due to the base branch is `remove-multiple-exports` and not staging, this will be merged into that branch and will run the visual tests there anyway.

braddialpad commented:
> Visual tests are marking all the snapshots as "new" due to the base branch is `remove-multiple-exports` and not staging, this will be merged into that branch and will run the visual tests there anyway.

ah yeah that makes sense

braddialpad commented:
> * Our current DtIcon implementation is not tree-shakeable and it will never be as long as were using a prop to determine which icon should render, we'll need to do a big migration if we want to be able to fully tree-shake our icons.

In theory if we were to use dynamic import for this wouldn't that fix the issue? Then the icon would only be imported when used?

juliodialpad commented:
> > * Our current DtIcon implementation is not tree-shakeable and it will never be as long as were using a prop to determine which icon should render, we'll need to do a big migration if we want to be able to fully tree-shake our icons.
> 
> In theory if we were to use dynamic import for this wouldn't that fix the issue? Then the icon would only be imported when used?

No, the way bundlers work is that if you do dynamic import it'd import every possibility so you're always importing all of them in that case.

We need to move out of dt-icon to make it fully tree-shakeable, another option would be to use CSS icons, something like `font awesome` were you only apply a class `<i class="fa fa-home" />` and the icons are rendered by CSS. That way, we would only need to use `dt-icon` to add that class dynamically.
